File: go/matching-brackets/matching_brackets.go
package brackets

var m = map[rune]rune{
	'(': ')',
	'[': ']',
	'{': '}',
}

func Bracket(input string) bool {

	var stack []rune

	for _, r := range input {

		for k, v := range m {
			if k == r {
				stack = append(stack, r)
				break
			} else if v == r {
				if len(stack) == 0 || stack[len(stack)-1] != k {
					return false
				}
				stack = stack[:len(stack)-1]
				break
			}
		}

	}

	return len(stack) == 0

}

File: clojure/matching-brackets/src/matching_brackets.clj
(ns matching-brackets)

(def ^:private matching {\{ \}, \[ \], \( \)})

(def ^:private opening? (set (keys matching)))

(def ^:private bracket? (into opening? (vals matching)))

(defn valid? [s]
  (let [update-stack (fn [acc c]
                       (cond (opening? c) (conj acc c)
                             (= (matching (peek acc)) c) (pop acc)
                             :else (reduced [::not-empty])))]
    (->> s
         (filter bracket?)
         (reduce update-stack [])
         (empty?))))
